source/array_list_with_map.c

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/allocator.h>
+#include <aws/common/array_list_with_map.h>
+#include <aws/common/device_random.h>
+#include <aws/common/string.h>
+
+int aws_array_list_with_map_init(
+    struct aws_array_list_with_map *list_with_map,
+    struct aws_allocator *allocator,
+    aws_hash_fn *hash_fn,
+    aws_hash_callback_eq_fn *equals_fn,
+    aws_hash_callback_destroy_fn *destroy_element_fn,
+    size_t item_size,
+    size_t initial_item_allocation) {
+    if (aws_array_list_init_dynamic(&list_with_map->list, allocator, initial_item_allocation, item_size)) {
+        return AWS_OP_ERR;
+    }
+
+    if (aws_hash_table_init(
+            &list_with_map->map, allocator, initial_item_allocation, hash_fn, equals_fn, destroy_element_fn, NULL)) {
+        aws_array_list_clean_up(&list_with_map->list);
+        return AWS_OP_ERR;
+    }
+    return AWS_OP_SUCCESS;
+}
+
+void aws_array_list_with_map_clean_up(struct aws_array_list_with_map *list_with_map) {
+    if (!list_with_map) {
+        return;
+    }
+    aws_array_list_clean_up(&list_with_map->list);
+    aws_hash_table_clean_up(&list_with_map->map);
+}
+
+int aws_array_list_with_map_insert(struct aws_array_list_with_map *list_with_map, const void *element) {
+    AWS_PRECONDITION(list_with_map);
+    AWS_PRECONDITION(element);
+    if (aws_array_list_with_map_exist(list_with_map, element)) {
```

Comment:
If it's "set something", yeah, it makes sense to "you didn't set it, but it's fine". But, if we insert something, "we didn't insert it for you, but it's fine??", it should just fail from my point of view.

Reply:
If the issue is the name, maybe we should rename from insert to something else. Looking at java for inspiration: https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#add(E)

Reply:
add and insert is bascially the same thing, right? If you didn't add something for real, but the add succeed. I think it's weird

Reply:
Insert implies to me an ordering, but sets don't have an order.

Reply:
Add is like make an element into a set, and then union.

Reply:
sure, rename it to add makes more sense, but I believe we should still error out for duplicates instead of something like was_created

Reply:
I disagree, I think trying to add the same element to a set twice is perfectly legitimate. We should only error out when something went wrong. @DmitriyMusatkin what do you think?

Reply:
add thing is not "set twice", that's not hash table. It failed to add a thing, so it's an error to me.

Reply:
Java, CPP, and Dotnet all have a boolean return value for set add type operations. None error out due to failure to add.
https://docs.oracle.com/javase/8/docs/api/java/util/Set.html#add-E-
https://www.cplusplus.com/reference/unordered_set/unordered_set/insert/
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add?view=net-6.0#system-collections-generic-hashset-1-add(-0)

Reply:
What do we expect the caller behavior to be? Do we expect it to be "oh no, everything is on fire fail hard" or "ugh, they can be equal sometimes, so let me just ignore the error". We should probably go with whatever makes user experience easier.

Set is a fairly common ds, and i think common mental model of everyone working with sets would be that it will ignore repeated insert, so leaning more towards option 2. 

On add vs insert, if im not mistaken its more of a legacy thing: c++ initially implemented set as ordered data structure backed by the tree, so it inherited the insert name from the tree. While java and other languages backed set using hash table and went with add